Tasks/Common/compression-common/FileExtractor.ts

```diff
@@ -0,0 +1,179 @@
+
+    // change the console output code page to UTF-8
+    const chcpPath = path.join(process.env.windir, "system32", "chcp.com");
+    await tl.exec(chcpPath, '65001');
+
+    // run powershell
+    const powershell: tr.ToolRunner = tl.tool('powershell')
+        .line('-NoLogo -Sta -NoProfile -NonInteractive -ExecutionPolicy Unrestricted -Command')
+        .arg(command);
+    await powershell.exec();
+}
+
+/** Extract a file using 7zip. */
+const sevenZipExtract = (isWindows: boolean) => async (file: string, destination: string): Promise<void> => {
+    console.log(tl.loc('SevenZipExtractFile', file));
+
+    const sevenZipLocation = isWindows ? path.join(__dirname, '7zip/7z.exe') : tl.which('7z', true);
+    const sevenZip = tl.tool(sevenZipLocation);
+
+    sevenZip.arg('x');
+    sevenZip.arg('-o' + destination);
+    sevenZip.arg(file);
+
+    const execResult = sevenZip.execSync();
+    if (execResult.code !== tl.TaskResult.Succeeded) {
+        tl.debug('execResult: ' + JSON.stringify(execResult));
+    }
+}
+
+/* Extract a compressed tar archive using the 'tar' utiltity. */
+async function tarExtract(file: string, destination: string): Promise<void> {
+    console.log(tl.loc('TarExtractFile', file));
+    const tr: tr.ToolRunner = tl.tool('tar');
+    tr.arg(['xzC', destination, '-f', file]);
+    tr.exec();
+}
+
+/** Choose an extractor function based on the file type and agent's operating system. */
+function pickExtractor(fileType: CompressedFile, hostOS: string): Extractor {
+    const isWindows = hostOS === 'win32';
+
+    if (isWindows) {
+        switch (fileType) {
+            case CompressedFile.Tar:
+            case CompressedFile.SevenZip:
+                return sevenZipExtract(isWindows);
+            case CompressedFile.Tarball:
+                return async (file: string, destination: string): Promise<void> => {
+                    // e.g. 'fullFilePath/test.tar.gz' --> 'test.tar.gz'
+                    const shortFileName = file.substring(file.lastIndexOf(path.sep) + 1, file.length);
+                    // e.g. 'destination/_test.tar.gz_'
+                    const tempFolder = path.normalize(path.join(destination, `_${shortFileName}_`));
+                    console.log(tl.loc('CreateTempDir', tempFolder, file));
+
+                    // create temp folder
+                    tl.mkdirP(tempFolder);
+
+                    // extract compressed tar
+                    await sevenZipExtract(isWindows)(file, tempFolder);
+                    console.log(tl.loc('TempDir', tempFolder));
+                    const tempTar = path.join(tempFolder, tl.ls('-A', [tempFolder])[0]); // should be only one
+                    console.log(tl.loc('DecompressedTempTar', file, tempTar));
+
+                    // expand extracted tar
+                    await sevenZipExtract(isWindows)(tempTar, destination);
+
+                    // clean up temp folder
```

Comment:
Tool installers must not just download and extract but it must prepend the path.  See the design doc, other tasks and the tool lib